### ISSUE TITLE
problems running rafdaf

### ISSUE THREAD

**Issue by janxkoci** [alanrogers/legofit#10]
Hi Alan, I'm trying to calculate derived allele frequencies for a set of populations and I'm having a problem with the `rafdaf` tool.

My script looks like this:

```bash
rafdaf \
       African=pops/African.raf \
       Sardinian=pops/Sardinian.raf \
       Papuan=pops/Papuan.raf \
       Altai=pops/Altai.raf \
       Vindija=pops/Vindija.raf \
       Chagyrskaya=pops/Chagyrskaya.raf \
       Denisova=pops/Denisova.raf \
       outgroup=pops/Apes.raf > archaics_african_sardinian_papuan.daf
```

Running the script produces this error.

```
$ bash rafdaf.sh 
error.c:mystrerror_r:55: buffer overflow
rafdaf.c:193: input error (Unknown error 1001)
```

I also got an output like this:

```
# rafdaf version 2.3.2-13-geb0240e
# Population labels:
#  African=pops/African.raf
#  Sardinian=pops/Sardinian.raf
#  Papuan=pops/Papuan.raf
#  Altai=pops/Altai.raf
#  Vindija=pops/Vindija.raf
#  Chagyrskaya=pops/Chagyrskaya.raf
#  Denisova=pops/Denisova.raf
#  outgroup=pops/Apes.raf
chr pos African Sardinian Papuan Altai Vindija Chagyrskaya Denisova
1 752721 0.19767442 0.94444444 0.1875 1 1 1 1
1 754163 0.97727273 0.94642857 0.94117647 1 1 1 1
```

So the tool runs, but there is some problem after the position **754163** at chromosome 1. So I had a look at the context after the position with `grep`:

```
grep -wHA 5 754163 pops/*raf > rafdaf_error_context.txt
```

The result shows that the next position present in all groups is **754618**:

```
$ grep 754618 rafdaf_error_context.txt 
pops/African.raf-1	754618	a	g	1
pops/Altai.raf-1	754618	a	g	1
pops/Apes.raf-1	754618	a	g	1
pops/Denisova.raf-1	754618	a	g	1
pops/Chagyrskaya.raf-1	754618	a	g	1
pops/Papuan.raf-1	754618	a	g	1
pops/Sardinian.raf-1	754618	a	g	1
pops/Vindija.raf-1	754618	a	g	1
```

Is this something that can cause a problem for `rafdaf`?

**Comment by alanrogers** [comment]
Thanks for pointing out this problem. I'll look into it.

It's been quite awhile since I used this program, because I usually let
sitepat call ancestral alleles instead. So it's possible that after making
some change to sitepat and tabpat, I failed to make the same change in
rafdaf. I'll let you know.

Out of curiosity, why are you using rafdaf? Are you using the results in
tabpat? (If so, try using sitepat instead.) Or are you using the output in
some non-Legofit program?

Alan

**Comment by alanrogers** [comment]
That was the problem. The function RAFReader_multiNext returns an integer, which rafdaf checks to make sure that everything is OK. One of the possible values is MONOMORPHIC_SITE. Rafdaf was failing to check for that value. It therefore aborted when it encountered a monomorphic site. Should be fixed now on the devlp branch, although not yet on master or main, or whatever I'm calling it in this repo.

**Comment by alanrogers** [comment]
Reopening this issue. I want to make sure the new code works for you before closing.

**Comment by janxkoci** [comment]
Thanks for a quick fix :) I will try to install the dev branch.

> Out of curiosity, why are you using rafdaf? Are you using the results in tabpat? (If so, try using sitepat instead.) Or are you using the output in some non-Legofit program?

I'm just trying to reproduce an analysis from one paper using our data, for which I need per-population derived allele frequencies. I would then process the results with custom scripts to look at what I need.

I've looked at [plink](https://www.cog-genomics.org/plink/1.9/basic_stats#freq)  and [admixtools2](https://uqrmaie1.github.io/admixtools/reference/plink_to_afs.html). Plink's output has a format that is kind of hard to work with, at first look, while admixtools2, being na R package, loads the data into memory (optionally in chunks), which is also a bit limiting. I even thought of writing something myself, at least as a validation check (and exercise in coding, which I like for the fun of it). Then I thought I could maybe use `raf` in some way to get what I need, and when I went to look at your docs, I've noticed `rafdaf`, which produces pretty much exactly what I need. The format is the same as from admixtools2, except I don't have to code myself the outgroup checking to figure out which allele is derived - your tool does it directly. And I prefer plaintext file directly over export from R, as an added bonus.
